### PR TITLE
Speedups and new vndf

### PR DESCRIPTION
**Pull request by tigrazone**

### CODE REVIEW

**Comment by tigrazone** [comment]
Also I chsnged ONB calculations to more precise. Read code please

File: src/HostDeviceCommon/Math.h [hunk 1]
```diff
@@ -80,12 +90,22 @@ namespace hippt
...
-#define M_PI 3.14159265358979323846f
+#define M_PI		3.14159265358979323846f
+#define M_TWO_PI	6.28318530717958647693f
+#define M_INV_PI	0.31830988618379067154f
+#define M_INV_2PI	0.63661977236758134308f
+#define M_TWO_PIPI	19.73920880217871723767f
+
+#define NEARzero	1.0e-10f
+
+	inline bool isZERO(float x) { return (x < NEARzero) && (x > -NEARzero); }
```
Comment (TomClabault) [inline src/HostDeviceCommon/Math.h:101]: Please rename `isZERO` to `is_zero()` for naming consistency. Also `NEARzero` --> `NEAR_ZERO`

File: src/Device/includes/ONB.h [hunk 2]
```diff
@@ -7,21 +7,39 @@
...
+ * and optimised little bit by @tigrazone
+*/
 HIPRT_HOST_DEVICE HIPRT_INLINE void build_ONB(const float3& N, float3& T, float3& B)
 {
-    float3 up = hippt::abs(N.z) < 0.9999999f ? make_float3(0.0f, 0.0f, 1.0f) : make_float3(1.0f, 0.0f, 0.0f);
-    T = hippt::normalize(hippt::cross(up, N));
-    B = hippt::cross(N, T);
+    if(N.z < -0.99998796F)  // Handle the singularity
+    {
+      T = make_float3(0.0F, -1.0F, 0.0F);
```
Comment (TomClabault) [inline src/Device/includes/ONB.h:19]: Same here

File: src/Device/includes/ONB.h [hunk 3]
```diff
@@ -7,21 +7,39 @@
...
+ * This uses the technique from "Improved accuracy when building an orthonormal basis" by Nelson Max, https://jcgt.org/published/0006/01/02.
+ * taken from https://github.com/nvpro-samples/nvpro_core/blob/master/nvvkhl/shaders/func.h
+ * and optimised little bit by @tigrazone
+*/
 HIPRT_HOST_DEVICE HIPRT_INLINE void build_ONB(const float3& N, float3& T, float3& B)
 {
-    float3 up = hippt::abs(N.z) < 0.9999999f ? make_float3(0.0f, 0.0f, 1.0f) : make_float3(1.0f, 0.0f, 0.0f);
-    T = hippt::normalize(hippt::cross(up, N));
-    B = hippt::cross(N, T);
+    if(N.z < -0.99998796F)  // Handle the singularity
```
Comment (TomClabault) [inline src/Device/includes/ONB.h:17]: `-1.0F` --> `-1.0f` with a lowercase F for consistency in the codebase.

File: src/Device/includes/ONB.h [hunk 4]
```diff
@@ -7,21 +7,39 @@
...
 
 /*
  * Rotation of the basis around the normal by 'basis_rotation' radians
  */
 HIPRT_HOST_DEVICE HIPRT_INLINE void build_rotated_ONB(const float3& N, float3& T, float3& B, float basis_rotation)
 {
-    float3 up = hippt::abs(N.z) < 0.9999999f ? make_float3(0.0f, 0.0f, 1.0f) : make_float3(1.0f, 0.0f, 0.0f);
-    T = hippt::normalize(hippt::cross(up, N));
+    if(N.z < -0.99998796F)  // Handle the singularity
+    {
```
Comment (TomClabault) [inline src/Device/includes/ONB.h:35]: Omit brackets with a single line in the if block

File: src/HostDeviceCommon/KernelOptions.h [hunk 5]
```diff
@@ -314,12 +314,14 @@
...
  *		caps as proposed in [Sampling Visible GGX Normals with Spherical Caps, Dupuy & Benyoub, 2023]
  * 
- *  - GGX_VNDF_BOUNDED [Not Yet Implemented]
+ *  - GGX_VNDF_BOUNDED
  *		Sample the distribution of visible normals with a bounded VNDF
  *		sampling range as proposed in [Bounded VNDF Sampling for Smith-GGX Reflections, Eto & Tokuyoshi, 2023]
  *		
  */
-#define GGXAnisotropicSampleFunction GGX_VNDF_SAMPLING
+//#define GGXAnisotropicSampleFunction GGX_VNDF_SAMPLING
```
Comment (TomClabault) [inline src/HostDeviceCommon/KernelOptions.h:322]: I think you can remove the commented lines here

File: src/HostDeviceCommon/KernelOptions.h [hunk 6]
```diff
@@ -314,12 +314,14 @@
...
+#define GGXAnisotropicSampleFunction GGX_VNDF_SPHERICAL_CAPS
+//#define GGXAnisotropicSampleFunction GGX_VNDF_BOUNDED
```
Comment (TomClabault) [inline src/HostDeviceCommon/KernelOptions.h]: And here

File: src/Device/includes/Sampling.h [hunk 7]
```diff
@@ -315,6 +316,40 @@ HIPRT_HOST_DEVICE HIPRT_INLINE float3 GGX_VNDF_spherical_caps_sample(const float
     return hippt::normalize(make_float3(alpha_x * Nh.x, alpha_y * Nh.y, Nh.z));
 }
 
+/**
+ * [Bounded VNDF Sampling for Smith–GGX Reflections, Kenta Eto and Yusuke Tokuyoshi; 2023]
+ */
+HIPRT_HOST_DEVICE HIPRT_INLINE float3 GGX_VNDF_bounded_sample(const float3 local_view_direction, float alpha_x, float alpha_y, Xorshift32Generator& random_number_generator)
```
Comment (TomClabault) [inline src/Device/includes/Sampling.h:322]: The overall implementation of the sampling function looks quite different from the reference which makes it harder to understand and follow.

I suppose this is all in the name of performance but in the grand scheme of things, simplifying a few float ops in a sampling function isn't worth it compared to the introduced divergence of the code from the reference in the paper.

I'd prefer an implementation closer to the one presented in the paper.

**Review by TomClabault — CHANGES_REQUESTED** [review]

**Comment by TomClabault** [comment]
Your implementation is also omitting the PDF (probability distribution function) change required for bounded VNDF sampling as presented in Listing 2 of [[Bounded VNDF Sampling for Smith–GGX Reflections]](https://gpuopen.com/download/publications/Bounded_VNDF_Sampling_for_Smith-GGX_Reflections.pdf).

This PDF change should be adapted in `disney_metallic_eval()`: https://github.com/TomClabault/HIPRT-Path-Tracer/blob/main/src/Device/includes/Disney.h#L102

File: src/HostDeviceCommon/KernelOptions.h (hunk 5 above)
Comment (tigrazone) [reply src/HostDeviceCommon/KernelOptions.h:322]: Commented for fast switch

**Pull request closed without merging**